customFieldDebugger.js

```diff
@@ -0,0 +1,407 @@
+}
+
+function enableTooltipVisualization() {
+  // Some code to render a basic hover tooltip UI element
+  document.querySelectorAll('[data-cfd-tooltip]').forEach(el => {
+    // Parse the data stored in the data-attributes cfdSource and cfdUsageType
+    let cfdInfo = {};
+    Object.entries(el.dataset).forEach(([key, val]) => {
+      if (key.includes('cfdSource')) {
+        const [_, id] = key.split('_');
+        if (!cfdInfo[id]) { cfdInfo[id] = {}; }
+        cfdInfo[id].source = el.dataset[key];
+      } else if (key.includes('cfdUsageType')) {
+        const [_, id] = key.split('_');
+        if (!cfdInfo[id]) { cfdInfo[id] = {}; }
+        cfdInfo[id].usage = el.dataset[key];
+      }
+    });
+
+    // Create & render a new div when the element is hovered
+    const toolTipEl = document.createElement('div');
+    toolTipEl.classList.add('CFD-tooltip');
+    let innerHTML = '';
+    Object.entries(cfdInfo).forEach(([key, val]) => {
+      innerHTML = innerHTML + `<div class="CFD-section"><span class="CFD-source">Data Field: ${val.source}</span><span class="CFD-usage">Usage: ${val.usage}</span></div>`;
+    });
+    toolTipEl.innerHTML = innerHTML;
+
+    el.addEventListener('mouseenter', () => {
+      const elRect = el.getBoundingClientRect();
+      toolTipEl.style.top = `${elRect.bottom + window.scrollY}px`;
+      toolTipEl.style.left = `${elRect.left}px`;
+      document.body.appendChild(toolTipEl);
+    });
+    el.addEventListener('mouseleave', () => {
+      toolTipEl.remove();
+    });
+  });
+}
+
+function enableUsageVisualization() {
+  const flatProfile = flattenObj(window._RSS_PROPS_NOPROXY_.data.document.streamOutput);
+
+  const cfEls = document.querySelectorAll('[data-cfd-tooltip]');
+  const usageMap = {};
+
+  Object.entries(flatProfile).forEach(([profileKey, _]) => {
+    // Search the page to see if this custom field is in use by checking the 'data-cfd-source' attribute
+    cfEls.forEach(el => {
+      Object.entries(el.dataset).forEach(([_, tooltipVal]) => {
+        if (profileKey === tooltipVal) {
+          usageMap[profileKey] = [
+            ...(usageMap[profileKey] ? usageMap[profileKey] : []),
+            el
+          ];
+        }
+      });
+    });
+  });
+
+  let usagePanelEl = document.createElement('div');
+  usagePanelEl.classList.add('CFD-usagePanel', 'CFDUsagePanel');
+  let usagePanelScrollableEl = document.createElement('div');
+  usagePanelScrollableEl.classList.add('CFDUsagePanel-scrollable');
+  usagePanelEl.appendChild(usagePanelScrollableEl);
+
+
+  function buildFieldSummaryHTML(key, val) {
+    let shortKey = key;
+    const prefix = 'document.streamOutput.';
+    if (shortKey.startsWith(prefix)) {
+      shortKey = shortKey.slice(prefix.length);
+    }
+
+    let shortVal = val;
+    const maxStrLen = 40;
+    if (typeof shortVal === 'string' && shortVal.length > maxStrLen) {
+      shortVal = `${shortVal.slice(0, maxStrLen)}...`;
+    }
+    return `
```

Comment:
I don't think we can avoid the dom manipulation stuff completely, but is it possible to implement the panel UI as a react component?